Improve Quickstart: Circuits (#2534)
PennyLaneAI/pennylane: doc/introduction/circuits.rst

    :target: javascript:void(0);

> In PennyLane, quantum computations are represented as *quantum node* objects. A quantum node or QNode is used to

Author: I took it out, it is actually not necessary here...

declare the quantum circuit, and also ties the computation to a specific device that executes it.

QNodes can interface with any of the supported numerical and machine learning libraries---:doc:`NumPy <interfaces/numpy>`,
:doc:`PyTorch <interfaces/torch>`, :doc:`TensorFlow <interfaces/tf>`, and
:doc:`JAX <interfaces/jax>`---indicated by providing an optional ``interface`` argument
when creating a QNode. Each interface allows the quantum circuit to integrate seamlessly with
library-specific data structures (e.g., NumPy and JAX arrays or Pytorch/TensorFlow tensors) and
:doc:`optimizers <optimizers>`.

> By default, QNodes use the NumPy interface. The other PennyLane interfaces are

Reviewer: it strikes me that nowhere in the intro mentions 'circuits'. I feel like we should, since a new reader that clicks on this page likely wants to know how this page relates to circuits?

Reviewer: I realize though that this is an existing problem with this page, separate to this PR!

Reviewer: Ah, the connection between circuits <--> qfuncs is a bit below, in the Quantum Functions section 🙂

Author: > In PennyLane, quantum computations, which involve the execution of one or more quantum circuits, 
are represented as *quantum node* objects.

What about this opener?

constraints:

* The quantum function accepts classical inputs, and consists of

>   :doc:`quantum operators <operations>` or sequences of operators called :doc:`templates`,

Reviewer: gosh the distinction between templates and operators is so arbitrary with the changes over the last year 😆

Author: Code-wise yes, but in research it is still very distinct :)

  :doc:`quantum operators <operations>` or sequences of operators called :doc:`templates`,
  using one instruction per line.

* The function can contain classical flow control structures such as ``for`` loops,
  but in general they must not depend on the parameters of the function.
* The function can contain classical flow control structures such as ``for``.

* The quantum function must always return either a single or a tuple of
  *measured observable values*, by applying a :doc:`measurement function <measurements>`
  to a :ref:`qubit observable <intro_ref_ops_qobs>` or :ref:`continuous-value observable <intro_ref_ops_cvobs>`.

.. note::

    Measured observables **must** come after all other operations at the end

>     of the circuit function as part of the return statement, and cannot appear in the middle.

Reviewer: oh gosh, nice catch!

    dev = qml.device('default.qubit', wires=2, shots=1000)

> PennyLane offers some basic devices such as the ``'default.qubit'``, ``'default.mixed'``
> and ``'default.gaussian'`` simulators; additional devices can be installed as plugins (see

Reviewer: ```suggestion
PennyLane offers some basic devices such as the ``'default.qubit'``, ``'default.mixed'``, ``lightning.qubit``,
and ``'default.gaussian'`` simulators; additional devices can be installed as plugins (see
```

Author: But lightning.qubit does not yet ship with PL? It is not in the `device` folder at least?

`available plugins <https://pennylane.ai/plugins.html>`_ for more details). Note that the
choice of a device significantly determines the speed of your computation, as well as
the available options that can be passed to the device loader.

> .. note::
> 
>     For example, check out the ``'lightning.qubit'`` `plugin <https://github.com/PennyLaneAI/pennylane-lightning>`_,
>     which is a fast state-vector simulator written in C++.

Reviewer: this is now a hard dependency if you install PL, so it comes across for free! I suggest maybe changing this to lightning.gpu?
```suggestion
.. note::

    For example, check out the ``'lightning.gpu'`` `plugin <https://github.com/PennyLaneAI/pennylane-lightning-gpu>`_,
    which is a fast state-vector simulator supporting GPUs.
```

Device options
^^^^^^^^^^^^^^

When loading a device, the name of the device must always be specified.

> Further options can then be passed as keyword arguments. These options can differ based
> on the device. For a plugin device, refer to the plugin documentation for available device options.

Reviewer: ```suggestion
Further options can then be passed as keyword arguments, and can differ based
on the device. For a plugin device, refer to the plugin documentation for available device options.
```

The two most important device options are the ``wires`` and ``shots`` arguments.

Wires

> *****

Reviewer: this section is so much clearer now!
